### PR TITLE
Drop puppet data

### PR DESCRIPTION
**Pull request by ezr-ondrej**
This adds a task to purge all puppet data, when plugin is not installed.
All the migrations will be owned by the plugin in the next version.

### CODE REVIEW

File: lib/tasks/cleanup.rake [hunk 1]
```diff
@@ -16,6 +16,37 @@ namespace :purge do
     Setting.where(name: 'max_trend').delete_all
   end
 
-  task all: ['purge:trends']
+  task puppet: :environment do
+    raise('You have Puppet plugin installed, uninstall it first to purge puppet data') if Foreman::Plugin.find(:foreman_puppet)
```
Comment (ekohl) [inline lib/tasks/cleanup.rake]:
```suggestion
    raise('You have the Puppet plugin installed, uninstall it first to purge puppet data') if Foreman::Plugin.find(:foreman_puppet)
```

File: db/migrate/20110616080444_add_look_up_key_id_to_puppet_class.rb [hunk 2]
```diff
@@ -1,7 +1,7 @@
 class AddLookUpKeyIdToPuppetClass < ActiveRecord::Migration[4.2]
   def up
+    # Will be dropped in later migration
```
Comment (ekohl) [inline db/migrate/20110616080444_add_look_up_key_id_to_puppet_class.rb:3]:
```suggestion
    # Will be dropped in a later migration
```

**Review by ekohl — COMMENTED** [review]
I still think it would be worth looking at #7961 as well.

Something else to consider: `db/migrate/20181224174419_add_index_to_environment_class_by_lookup_key_and_puppetclass.rb` is the newest migration file and it was introduced as part of 1.21.0. That means DB migrations from < 1.21.0 will be so unreliable that I'd call it unsupported. That means we should modify our upgrade test here: https://github.com/theforeman/jenkins-jobs/blob/1a91d93a7ee9103655233e90df242bde37cfedfb/theforeman.org/yaml/jobs/test_develop_pr_upgrade.yaml#L36. For those who don't want to decipher it: we have an upgrade job in every PR test that installs 1.17 and tests whether it can upgrade it to the latest code.

**Comment by ezr-ondrej** [comment]
[test katello]

File: lib/tasks/cleanup.rake [hunk 3]
```diff
@@ -16,6 +16,37 @@ namespace :purge do
     Setting.where(name: 'max_trend').delete_all
   end
 
-  task all: ['purge:trends']
+  task puppet: :environment do
+    raise('You have the Puppet plugin installed, uninstall it first to purge puppet data') if Foreman::Plugin.find(:foreman_puppet)
+
+    ActiveRecord::Base.connection.remove_referce(:template_combinations, :environment) if column_exists?(:template_combinations, :environment_id)
```
Comment (ekohl) [inline lib/tasks/cleanup.rake]: Is this a typo and should it be `remove_reference`?

**Review by ares — CHANGES_REQUESTED** [review]
There's a problem with the cleanup rake task. After I removed the plugin rpm, I got 

```
foreman-rake purge:puppet --trace
** Invoke purge:puppet (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute purge:puppet
rake aborted!
NoMethodError: undefined method `column_exists?' for main:Object
/usr/share/foreman/lib/tasks/cleanup.rake:22:in `block (2 levels) in <top (required)>'
/usr/share/gems/gems/rake-12.3.3/lib/rake/task.rb:273:in `block in execute'
/usr/share/gems/gems/rake-12.3.3/lib/rake/task.rb:273:in `each'
/usr/share/gems/gems/rake-12.3.3/lib/rake/task.rb:273:in `execute'
/usr/share/gems/gems/rake-12.3.3/lib/rake/task.rb:214:in `block in invoke_with_call_chain'
/usr/share/ruby/monitor.rb:235:in `mon_synchronize'
/usr/share/gems/gems/rake-12.3.3/lib/rake/task.rb:194:in `invoke_with_call_chain'
/usr/share/gems/gems/rake-12.3.3/lib/rake/task.rb:183:in `invoke'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:160:in `invoke_task'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:116:in `each'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:116:in `block in top_level'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:125:in `run_with_threads'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:110:in `top_level'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:83:in `block in run'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:186:in `standard_exception_handling'
/usr/share/gems/gems/rake-12.3.3/lib/rake/application.rb:80:in `run'
/usr/share/gems/gems/rake-12.3.3/exe/rake:27:in `<top (required)>'
/usr/bin/rake:23:in `load'
/usr/bin/rake:23:in `<main>'
Tasks: TOP => purge:puppet
```

Once this is fixed, I'm happy to retest (please ping me directly). Btw is there some write up on what all to perform for the successful disabling of puppet?

This is what I thought about so far

```
yum remove rubygem-foreman_puppet
foreman-rake purge:puppet
foreman-installer --no-enable-puppet
yum remove puppetserver
```
EDIT[Ondrej]: foreman-installer --no-enable-puppet --no-enable-foreman-puppet

**Comment by ekohl** [comment]
> This is what I thought about so far
> 
> ```
> yum remove rubygem-foreman_puppet
> foreman-rake purge:puppet
> foreman-installer --no-enable-puppet
> yum remove puppetserver
> ```
> 
> EDIT[Ondrej]: foreman-installer --no-enable-puppet --no-enable-foreman-puppet

If you have a Foreman Proxy also `--foreman-proxy-puppet false --foreman-proxy-puppet-ca false`.
If you have hammer also `--no-enable-foreman-cli-puppet` (and `yum remove rubygem-hammer_cli_foreman_puppet`)

And on katello/foreman-proxy-content scenarios also `--foreman-proxy-content-puppet false`

**Comment by ares** [comment]
and `systemctl foreman restart` (since the plugin was removed), I'll test the final installer command once the data cleanup is complete

**Comment by ares** [comment]
As discussed on IRC we should also clean environment permissions and Puppet feature from proxies, then it's ready.

**Review by ares — APPROVED** [review]
Works great now! The foreman test is unrelated (though I haven't seen this before). Waiting on katello tests, then happy to merge!

**Comment by ares** [comment]
just leaving a note for 3.0 - `--foreman-proxy-puppet-ca false` is not available and the same for `--no-enable-puppet` but of course we need instructions for 3.1

**Comment by ekohl** [comment]
It's `--foreman-proxy-puppetca false`. I shouldn't try to write that from the top of my head without verification.

And `--no-enable-puppet` should always be available. That's weird. What does `--help` or `--full-help` show?

File: db/migrate/20201013191730_drop_environment_from_host_and_hostgroup.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    if ActiveRecord::Base.connection.column_exists?(:hosts, :environment_id)
+      remove_foreign_key :hosts, :environments, name: 'hosts_environment_id_fk'
+      remove_column :hosts, :environment_id
+    end
+    if ActiveRecord::Base.connection.column_exists?(:hostgroups, :environment_id)
+      remove_foreign_key :hostgroups, :environments, name: 'hostgroups_environment_id_fk'
+      remove_column :hostgroups, :environment_id
+    end
+  end
+end
```
Comment (tbrisker) [inline db/migrate/20201013191730_drop_environment_from_host_and_hostgroup.rb:12]: A bit late to the party, but this is dangerous. If a user upgrades foreman without installing the plugin (e.g. when using a manual instalaltion without the installer) they will lose the environment data before it is migrated to the facet. This should be done inside the cleanup rake task, not in a migration.

Comment (tbrisker) [reply db/migrate/20201013191730_drop_environment_from_host_and_hostgroup.rb:12]: https://github.com/theforeman/foreman/pull/8849

**Pull request merged**